File: backend/app/timeline.py
from __future__ import annotations

import json
from typing import Any


def create_part(
    part_id: str, kind: str, *, status: str = "running", **fields: Any
) -> dict[str, Any]:
    return {"id": part_id, "kind": kind, "status": status, **fields}


def append_text(part: dict[str, Any], delta: str) -> dict[str, Any]:
    next_part = dict(part)
    next_part["text"] = f"{part.get('text', '')}{delta}"
    return next_part


def finalize_part(part: dict[str, Any]) -> dict[str, Any]:
    next_part = dict(part)
    next_part["status"] = "done"
    return next_part


def fail_part(part: dict[str, Any], detail: str) -> dict[str, Any]:
    next_part = dict(part)
    next_part["status"] = "error"
    next_part["detail"] = detail
    return next_part


def serialize_parts(parts: list[dict[str, Any]]) -> str:
    return json.dumps(parts, ensure_ascii=False)


def legacy_message_parts(content: Any, thinking_text: str) -> list[dict[str, Any]]:
    text_content = content if isinstance(content, str) else ""
    parts: list[dict[str, Any]] = []
    if thinking_text:
        parts.append(
            create_part(
                "legacy-thinking",
                "thinking",
                status="done",
                text=thinking_text,
            )
        )
    if text_content:
        parts.append(
            create_part(
                "legacy-answer",
                "answer",
                status="done",
                text=text_content,
            )
        )
    return parts


def message_parts_from_row(row: Any) -> list[dict[str, Any]]:
    if row["role"] != "assistant":
        return []
    if row["content_json"]:
        content = json.loads(row["content_json"])
        if isinstance(content, dict) and isinstance(content.get("parts"), list):
            return content["parts"]
    return legacy_message_parts(row["content_text"] or "", row["thinking_text"] or "")


def answer_text_from_parts(parts: list[dict[str, Any]]) -> str:
    return "".join(
        part.get("text", "")
        for part in parts
        if isinstance(part, dict) and part.get("kind") == "answer"
    )


def thinking_text_from_parts(parts: list[dict[str, Any]]) -> str:
    return "".join(
        part.get("text", "")
        for part in parts
        if isinstance(part, dict) and part.get("kind") == "thinking"
    )


def assistant_content_from_row(row: Any) -> str:
    parts = message_parts_from_row(row)
    answer_text = answer_text_from_parts(parts)
    if answer_text:
        return answer_text
    content_text = row["content_text"]
    return content_text if isinstance(content_text, str) else ""

File: frontend/src/components/chat/timeline.ts
import type { ChatStreamEvent, Message, TimelinePart } from '../../types'

export type TimelineState = {
  parts: TimelinePart[]
  expandedById: Record<string, boolean>
  manuallyExpanded: Record<string, boolean>
}

export function toRenderableTimeline(message: Message): TimelinePart[] {
  if (Array.isArray(message.parts) && message.parts.length > 0) {
    return message.parts
  }

  if (
    typeof message.content === 'object' &&
    message.content !== null &&
    !Array.isArray(message.content) &&
    Array.isArray(message.content.parts)
  ) {
    return message.content.parts
  }

  const parts: TimelinePart[] = []

  if (message.thinking_text) {
    parts.push({
      id: `legacy-thinking-${message.id}`,
      kind: 'thinking',
      status: 'done',
      text: message.thinking_text,
    })
  }

  if (typeof message.content === 'string' && message.content) {
    parts.push({
      id: `legacy-answer-${message.id}`,
      kind: 'answer',
      status: 'done',
      text: message.content,
    })
  }

  return parts
}

export function applyTimelineEvent(state: TimelineState, event: ChatStreamEvent): TimelineState {
  if (event.type === 'timeline_part_start') {
    const nextExpandedById = Object.fromEntries(
      state.parts.map((part) => [
        part.id,
        Boolean(state.manuallyExpanded[part.id] || state.expandedById[part.id]),
      ]),
    ) as Record<string, boolean>

    return {
      parts: [...state.parts, event.part],
      expandedById: { ...nextExpandedById, [event.part.id]: true },
      manuallyExpanded: state.manuallyExpanded,
    }
  }

  if (event.type === 'timeline_part_delta') {
    return {
      ...state,
      parts: state.parts.map((part) => {
        if (part.id !== event.part_id) return part
        const nextPart = { ...part, ...event.delta }
        if (typeof event.delta.text === 'string') {
          nextPart.text = `${part.text ?? ''}${event.delta.text}`
        }
        return nextPart
      }),
    }
  }

  if (event.type === 'timeline_part_end') {
    return {
      ...state,
      parts: state.parts.map((part) => (part.id === event.part_id ? { ...part, status: 'done' } : part)),
    }
  }

  if (event.type === 'timeline_part_error') {
    return {
      ...state,
      parts: state.parts.map((part) =>
        part.id === event.part_id ? { ...part, status: 'error', detail: event.detail } : part,
      ),
      expandedById: { ...state.expandedById, [event.part_id]: true },
    }
  }

  return state
}
